subprojects/scala/src/main/java/org/gradle/api/tasks/scala/ScalaDoc.java

```diff
@@ -197,7 +199,13 @@ protected void generate() {
             parameters.getOptionsFile().set(optionsFile);
             parameters.getClasspath().from(getClasspath());
             parameters.getOutputDirectory().set(getDestinationDir());
-            parameters.getSources().from(getSource());
+            boolean isScala3 = ScalaRuntimeHelper.findScalaJar(getScalaClasspath(), "library_3") != null;
+            parameters.getIsScala3().set(isScala3);
+            if (isScala3) {
```

Comment:
This effectively assumes that a custom `ScalaDoc` task on a custom source set would properly wire both the `classpath` and the `sources` and that we can simply here change what is to be considered "sources"